[Flang][OpenMP] Implement device clause lowering for target directive

chichunchen opened the pull request:
Add lowering support for the OpenMP `device` clause on the `target` directive in Flang.

The device expression is propagated through MLIR OpenMP and passed to the host-side `__tgt_target_kernel` call.

chichunchen committed:
[Flang][OpenMP] Implement device clause lowering for target directive

Add lowering support for the OpenMP `device` clause on the `target`
directive in Flang.

The device expression is propagated through MLIR OpenMP and passed to
the host-side `__tgt_target_kernel` call.

chichunchen committed:
Embedding deviceID inside runtimeAttr instead of changing createTarget interface

chichunchen committed:
Apply clang-format

ergawy reviewed (commented):
Thanks! Just one small comment about proberly reporting current lowering status.

tblah reviewed (commented):
Thank you for implementing this. Please could you update flang/docs/OpenMPSupport.md

chichunchen committed:
Resolve feedback

- Update OpenMP support for the device clause on the target construct
- Remove checkDevice from the target construct.
- Add checkDevice on target_update and target_data
- Does not add checkDevice on target_enter and target_exit
    - Adding checkDevice to target_enter and target_exit would cause a
      regression in omptarget-llvm.mlir, since that test currently uses
      the device clause in both target_enter and target_exit but expects
      -1 to be passed to the __tgt_target_data* APIs.

chichunchen commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -340,7 +340,7 @@ static LogicalResult checkImplementationStatus(Operation &op) {
   };
   auto checkDevice = [&todo](auto op, LogicalResult &result) {
-    if (op.getDevice())
+    if (op.getDevice() && !isa<omp::TargetOp>(op))
```
I removed checkDevice from TargetOp and added it to TargetUpdateOp and TargetDataOp. I did not add checkDevice to TargetEnterDataOp or TargetExitDataOp, because doing so would cause regressions in [omptarget-llvm.mlir](https://github.hpe.com/hpcde/cce-cce/blame/master/mlir/test/Target/LLVMIR/omptarget-llvm.mlir)￼. That test uses the device clause but hardcodes the device_id to -1 when calling the __tgt_target_enter_data and __tgt_target_exit_data APIs.

ergawy reviewed (approved):
LGTM! Thanks

tblah reviewed (approved):
Thanks!
